### ISSUE TITLE
About the implementation

### ISSUE THREAD

**Issue by boluoweifenda** [adobe/antialiased-cnns#7]
Thanks for your great work! I have two questions regarding the implementation details:
(1) In the situation of strided-convolution, why the BlurPool layer is placed after the ReLU rather than right next to the convolution?
It would be much more flexible if the conv and blurpool can be coupled.
I was considering the implementation in the pre-activation resnet.
(2) This question might be silly, but why not apply bilinear interpolation layers to downsample the feature map? I haven't seen any work use it.

**Comment by richzhang** [comment]
Thanks for the questions.

(1) If we do `Conv(s1)-BlurPool(s2)-Relu`, that is equivalent to `Blur(s1)-Conv(s2)-Relu`, since `Blur` and `Conv` are commutable. This is equivalent to blurring the input before processing it, destroying information unnecessarily. Performance will go down.

(2) Yes, using [1 2 1] is essentially equivalent to bilinear interpolation

**Comment by boluoweifenda** [comment]
Thanks for your quick reply, I will rethink deeply into these.

**Comment by wandering007** [comment]
> If we do Conv(s1)-BlurPool(s2)-Relu, that is equivalent to Blur(s1)-Conv(s2)-Relu, since Blur and Conv are commutable. 

I don't think these two are strictly equivalent. Here is the code snippet to verify it:

```
import torch
import torch.nn as nn

conv = nn.Conv2d(3, 3, 3, stride=1, padding=1)

def test_conv_blur():
    inputs = torch.randn(1, 3, 32, 32)
    blur = nn.AvgPool2d(2, 2)
    with torch.no_grad():
        outputs1 = conv(blur(inputs))
        outputs2 = blur(conv(inputs))
    assert torch.equal(outputs1, outputs2), "diff sum: {}".format(torch.sum(torch.abs(outputs1 - outputs2)).sum())

if __name__ == "__main__":
    test_conv_blur()
```

The assert cannot be passed. 

Furthermore, the increased FLOPs might contribute to the increased accuracy.

**Comment by richzhang** [comment]
Thanks for the code snippet. Your implementation differs from my statement in a critical way -- note the striding.

The readme has a new plot showing accuracy vs run-time.

**Comment by wandering007** [comment]
Thanks for your quick reply and pointing out my mistakes. I update the code snippet but the assert still cannot be passed.

**Comment by richzhang** [comment]
Maybe you can print the norm of the error vs the norm of the output signal. I suspect the discrepancy is due to numerical issues.

In any case, the equivalence should be provable. The avgpool is a convolution with a [1 1; 1 1] filter, and convolution is a commutable operation. In fact, you could combine the two operations by applying [1 1; 1 1] to the 3x3 conv kernels, making a single 4x4 conv layer.

**Comment by wandering007** [comment]
The average difference value is about 0.01, which cannot be caused by numerical issues.

**Comment by richzhang** [comment]
Maybe you can take some time to go through the proof that convolutions are associative and communative. https://en.wikipedia.org/wiki/Convolution

```
import torch
import torch.nn as nn

inp = torch.randn(1, 3, 32, 32)
conv = nn.Conv2d(3, 3, 3, stride=1, padding=0)
blur = nn.AvgPool2d(3, stride=1, padding=0)

out1 = conv(blur(inp))
out2 = blur(conv(inp))
```

(1.284619802805647e-15, 0.05549472197890282)

**Comment by wandering007** [comment]
Thanks. I just figure it out. `padding` matters a lot. I set `padding=1`. Extra zeros around the image border cause the inequality.
